pkg/kaniko/context_remote.go

```diff
@@ -54,29 +56,81 @@ func (c RemoteContextProvider) PrepareContext(opts types.ImageBuilderOpts) (stri
 
-	tarGzArchiver := archiver.TarGz{
-		Tar: &archiver.Tar{
-			OverwriteExisting:      true,
-			MkdirAll:               true,
-			ImplicitTopLevelFolder: false,
-			ContinueOnError:        false,
-		},
-		CompressionLevel: gzip.BestCompression,
+	// Check if the build context directory exists.
+	if _, err := os.Stat(buildContextDir); os.IsNotExist(err) {
+		return fmt.Errorf("can't access directory %q: it doesn't exist", buildContextDir)
 	}
 
-	var filesToArchive []string
+	// Walk through the build context directory, and collect all the files to be archived.
+	files := make(map[string]os.FileInfo)
+	if err := filepath.Walk(buildContextDir, func(filePath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("error accessing file %s: %w", filePath, err)
+		}
 
-	fileOrDirInfos, err := os.ReadDir(buildContextDir)
-	if err != nil {
-		return fmt.Errorf("can't access directory %s, err is : %w", buildContextDir, err)
+		if !info.IsDir() {
+			files[filePath] = info
+		}
+
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error walking the build context directory %s: %w", buildContextDir, err)
 	}
 
-	for _, fileOrDir := range fileOrDirInfos {
-		filesToArchive = append(filesToArchive, path.Join(buildContextDir, fileOrDir.Name()))
+	// Create the directory for the .tar.gz file if it doesn't exist.
+	if err := os.MkdirAll(path.Dir(tarGzPath), 0o750); err != nil {
+		return fmt.Errorf("can't create the archive destination directory %q: %w", path.Dir(tarGzPath), err)
+	}
+
+	// Create the .tar.gz file.
+	tarGzFile, err := os.Create(tarGzPath)
+	if err != nil {
+		return fmt.Errorf("can't create tar.gz file %s: %w", tarGzPath, err)
 	}
+	defer func() {
+		_ = tarGzFile.Close()
+	}()
+
+	gzipWriter := gzip.NewWriter(tarGzFile)
+	defer func() {
+		_ = gzipWriter.Close()
+	}()
+
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer func() {
+		_ = tarWriter.Close()
+	}()
+
+	for filePath, info := range files {
+		// Create a tar header for the file.
+		header, err := tar.FileInfoHeader(info, filePath)
+		if err != nil {
+			return fmt.Errorf("error creating tar header for file %s: %w", filePath, err)
+		}
+
+		// Update the header name to be relative to the build context directory
+		relPath, err := filepath.Rel(buildContextDir, filePath)
+		if err != nil {
+			return fmt.Errorf("error getting relative path for file %s: %w", filePath, err)
+		}
+		header.Name = relPath
+
+		// Write the header to the tar archive.
+		if err := tarWriter.WriteHeader(header); err != nil {
+			return fmt.Errorf("error writing tar header for file %s: %w", header.Name, err)
+		}
+
+		// Open the file and write its contents to the tar archive.
+		file, err := os.Open(filePath)
+		if err != nil {
+			return fmt.Errorf("error opening file %s: %w", filePath, err)
+		}
+
+		if _, err := io.Copy(tarWriter, file); err != nil {
+			return fmt.Errorf("error writing file %s to tar archive: %w", filePath, err)
+		}
```

Comment:
I think we can create a helper for tar write

Reply:
What do you mean by a helper?

Reply:
Ok, done 👌